src/main/java/com/es/phoneshop/cart/HttpSessionCartService.java

```diff
@@ -0,0 +1,55 @@
+package com.es.phoneshop.cart;
+
+
+import com.es.phoneshop.exceptions.OutOfStockException;
+import com.es.phoneshop.model.product.ArrayListProductDao;
+import com.es.phoneshop.model.product.Product;
+
+import javax.servlet.http.HttpServletRequest;
+import java.math.BigDecimal;
+import java.util.Optional;
+
+public class HttpSessionCartService implements CartService {
+    protected static final String CART_SESSION_ATTRIBUTE = HttpSessionCartService.class + ".cart";
+    private static HttpSessionCartService httpSessionCartService = new HttpSessionCartService();
+    private ArrayListProductDao productDao;
+
+    private HttpSessionCartService() {
+        productDao = ArrayListProductDao.getInstance();
+    }
+
+    public static HttpSessionCartService getInstance() {
+        return httpSessionCartService;
+    }
+
+    @Override
+    public Cart getCart(HttpServletRequest request) {
+        Cart result = (Cart) request.getSession().getAttribute(CART_SESSION_ATTRIBUTE);
+        if (result == null) {
+            result = new Cart();
+            request.getSession().setAttribute(CART_SESSION_ATTRIBUTE, result);
+        }
+        return result;
+    }
+
+    @Override
+    public void add(Cart cart, Long productId, Long quantity) throws OutOfStockException {
+        Product product = productDao.getProduct(productId).get();
+        if (quantity > product.getStock() || quantity <= 0) {
+            throw new OutOfStockException(product.getStock());
+        }
+        Optional<CartItem> optionalCartItem = cart.getCartItems()
+                .stream()
+                .filter(cartItem -> cartItem.getProductId().equals(productId))
+                .findFirst();
+        if (optionalCartItem.isPresent()) {
+            CartItem cartItem = optionalCartItem.get();
+            cartItem.setQuantity(cartItem.getQuantity() + quantity);
```

Comment:
What if the stock for a product is equal to 5 and I will add this product to a cart 10 times (always specifying a quantity of 1)?